service/storage_service.cc

```diff
@@ -1158,6 +1158,46 @@ class topology_coordinator {
         }
     }
 
+    // Preconditions:
+    // - There are no nodes that are trying to join and there are no topology
+    //   operations in progress
```

Comment:
Why do we care that there other nodes that are trying to join?

Reply:
> It's not important yet without the new handshake, so maybe it was added prematurely.

On the contrary. With new handshake we do not need to wait for all nodes that want to be joined.

> 
> > We will check their features when we will try to join them.
> 
> We need to make this procedure retryable. The topology coordinator needs to:
> 
>     1. Check the features,
> 
>     2. Respond to the node with the decision,
> 
>     3. Apply a command to change the state of the topology SM.
> 
> 
> In particular, the step 3. can fail (as well as 2., but not important here), so if there is a joining node then we can't be sure whether we responded to them or not yet. Therefore, we shouldn't enable features between retries because it could affect our decision.

We do raft barrier before retries (taking a guard). We know exactly if previous operation succeeded or not.
But even if not it is fine to not enable features after retrying a failed attempt, but there is no reason to not enable features after a successful bootstrap complete and before next one is started.